javaslang/src/main/java/javaslang/$.java

```diff
@@ -33,8 +33,7 @@
     // -- javaslang.collection
 
     // List
-    @Unapply static <T> Tuple2<T, List<T>> List(List.Cons<T> cons) { return Tuple.of(cons.head(), cons.tail()); }
-    @Unapply static <T> Tuple0 List(List.Nil<T> nil) { return Tuple.empty(); }
+    @Unapply static <T> Tuple2<T, List<T>> List(List<T> cons) { return Tuple.of(cons.head(), cons.tail()); }
```

Comment:
We need them in order to decompose algebraic data types (ADTs) (see my [previous comment](https://github.com/javaslang/javaslang/pull/1323/files/0756c6164f8e6c186578cee0cbecf2ee101e43a8#r62576805)).

Your version does throw `NoSuchElementException` on empty List (see `cons.head()`):

```java
@Unapply static <T> Tuple2<T, List<T>> List(List<T> cons) { return Tuple.of(cons.head(), cons.tail()); }
```

`Option` is a good example for an ADT, it can be `Some` and `None`. `List` is another example for an ADT, it can be `Cons` and `Nil`. Because `Stream` is similar to `List`, it can be `Cons` and `Empty`, I decided to choose different names than in Scala. In Javaslang it is just `List` and `Stream`.

Scala:

```scala
list match {
    case head :: tail => ...
    case Nil => ...
}
```

Javaslang:

```java
Match(list).of(
    Case(List($(), $()), (head, tail) => ...),
    Case(List(), ...)
);
```

We can't generalize the pattern matching framework to decompose arbitrary types into head/tail because we decide it _by type_  and not by _isEmpty_ check.